server/sonar-server/src/main/java/org/sonar/server/user/ServerUserSession.java

```diff
@@ -136,33 +134,27 @@ public boolean hasOrganizationPermission(String organizationUuid, String permiss
   }
 
   @Override
-  public boolean hasComponentUuidPermission(String permission, String componentUuid) {
-    if (isRoot()) {
-      return true;
-    }
-
+  protected Optional<String> componentUuidToProjectUuid(String componentUuid) {
     String projectUuid = projectUuidByComponentUuid.get(componentUuid);
-    if (projectUuid == null) {
-      try (DbSession dbSession = dbClient.openSession(false)) {
-        Optional<ComponentDto> component = dbClient.componentDao().selectByUuid(dbSession, componentUuid);
-        if (!component.isPresent()) {
-          return false;
-        }
-        projectUuid = component.get().projectUuid();
-      }
+    if (projectUuid != null) {
+      return Optional.of(projectUuid);
     }
-    boolean hasComponentPermission = hasProjectPermissionByUuid(permission, projectUuid);
-    if (hasComponentPermission) {
+    try (DbSession dbSession = dbClient.openSession(false)) {
+      com.google.common.base.Optional<ComponentDto> component = dbClient.componentDao().selectByUuid(dbSession, componentUuid);
+      if (!component.isPresent()) {
+        return Optional.empty();
+      }
+      projectUuid = component.get().projectUuid();
       projectUuidByComponentUuid.put(componentUuid, projectUuid);
-      return true;
+      return Optional.of(projectUuid);
     }
-    return false;
   }
 
-  // To keep private
-  private boolean hasProjectPermissionByUuid(String permission, String projectUuid) {
+  @Override
+  protected boolean hasProjectUuidPermission(String permission, String projectUuid) {
     if (!projectPermissionsCheckedByUuid.contains(permission)) {
       try (DbSession dbSession = dbClient.openSession(false)) {
+        // FIXME do not load all the authorized projects. It can be huge.
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Take the required action to fix the issue indicated by this comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1134)